Vadinur/profile 1 3

vadimrn opened the pull request.

vadimrn committed:
Декомпозиция профиля, добавлен файл md.

vadimrn committed:
редактирование AppDelegate

vadimrn committed:
Создание UITabBarController, который управляет несколькими контроллерами представления через панель вкладок в нижней части экрана. На архитектуре MVVM

vadimrn committed:
создание контроллера представления WebViewViewController

vadimrn committed:
ProfileRouter отвечает за навигацию

vadimrn committed:
Добавление ProfileViewModel, отвечает за управление и обработку данных профиля пользователя

vadimrn committed:
UserProfileModel представляет данные профиля пользователя, а ProfileModel отвечает за взаимодействие с сетевым слоем для получения и обновления этих данных.

vadimrn committed:
контроллер редактирования профиля пользователя

vadimrn committed:
Контроллер редактирования профиля пользователя

vadimrn committed:
заглушка

vadimrn committed:
заглушка

vadimrn committed:
заглушка корзины

vadimrn committed:
Заглушка статистики

vadimrn committed:
добавлены цвета

vadimrn committed:
цвета

vadimrn committed:
стили шрифтов

vadimrn committed:
скрытием клавиатуры

vadimrn committed:
добавление пользовательских цветов

vadimrn committed:
шрифты

vadimrn committed:
создание observable

vadimrn committed:
создание и настройка контроллеров

vadimrn committed:
(alerts) в пользовательском интерфейсе

vadimrn committed:
создание и настройка элементов пользовательского интерфейса

vadimrn committed:
механизм для проверки

vadimrn committed:
commit

vadimrn committed:
донастройка

vadimrn committed:
доработки

vadimrn committed:
(перезалил файлы из своих черновиков)

vadimrn committed:
декомпозиция на доске

DanilaG commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,125 @@
+            ProgressHUD.dismiss()
+            switch result {
+            case .success(let userProfile):
+                self.userProfile = userProfile
```
userProfile не является thread safety, поэтому к нему следует обращаться только из одного потока. Предлагаяю обращаться только из главного

vadimrn replied:
исправил

DanilaG commented on FakeNFT/Profile/ProfileScreen/Views/ProfileViewModel.swift:
```diff
@@ -0,0 +1,125 @@
+            guard let self = self else { return }
+            switch result {
+            case .success(let updatedProfile):
+                self.userProfile = updatedProfile
```
Непотокобезопастно, так как доступ к userProfile идёт из разных потоков

vadimrn replied:
исправил

DanilaG commented on FakeNFT/Helpers/ViewFactory.swift:
```diff
@@ -0,0 +1,19 @@
+        textView.font = .bodyRegular
+        textView.backgroundColor = .nftLightgrey
+        textView.layer.cornerRadius = 12
+        textView.textContainerInset = UIEdgeInsets(top: 11, left: 10, bottom: 11, right: 10)
```
Горизонтальные отступы по 16 в макетах

vadimrn replied:
исправил

DanilaG commented on FakeNFT/DesignSystem/Fonts.swift:
```diff
@@ -1,19 +1,21 @@
-    static var caption2 = UIFont.systemFont(ofSize: 13, weight: .regular)
+    static var caption1 = UIFont(name: nftFontRegularName, size: 15)
+    static var caption2 = UIFont(name: nftFontRegularName, size: 13)
+    static var caption3 = UIFont(name: nftFontMediumName, size: 10)
```
Давай поправим на let, чтобы случайно не изменить шрифты

vadimrn replied:
Исправил, заменено на let

DanilaG commented on FakeNFT/Profile/ProfileScreen/Views/ProfileViewController.swift:
```diff
@@ -0,0 +1,236 @@
+                        self?.userNameLabel.text = model.name
+                        self?.userDescriptionLabel.text = model.description
+                        self?.userWebSiteTextView.text = model.website
+                        [self!.editButton, self!.profileImageView, self!.userNameLabel, self!.userDescriptionLabel, self!.userWebSiteTextView, self!.profileTableView].forEach { $0.isHidden = false }
```
Небезопасная работа с force unwrap, приведёт к падению, если пользователь выйдет с экрана, но setImage вернёт результат

vadimrn replied:
Исправил

DanilaG commented on FakeNFT/Profile/ProfileScreen/Views/ProfileViewController.swift:
```diff
@@ -0,0 +1,236 @@
+                        self?.userNameLabel.text = model.name
+                        self?.userDescriptionLabel.text = model.description
+                        self?.userWebSiteTextView.text = model.website
+                        [self!.editButton, self!.profileImageView, self!.userNameLabel, self!.userDescriptionLabel, self!.userWebSiteTextView, self!.profileTableView].forEach { $0.isHidden = false }
```
Давай ограничем длину строки, например до 100 или 120 символов, что позволит сделать код более читаемым

vadimrn committed:
Исправления согласно 2-го ревью. Горизонтальные отступы по 16 в макетах

vadimrn committed:
Исправлено на let, чтобы не изменились шрифты

vadimrn committed:
обернут доступ к userProfile в GCD-блок

vadimrn committed:
создана ветка ProfileUI+BascicLogic

vadimrn committed:
Merge branch 'vadinur/profile-1-3' of https://github.com/Kolesnikov-Eugene/iOS-FakeNFT-StarterProject-Public into vadinur/profile-1-3

vadimrn committed:
Обеспечение потокобезопасности доступа к UserProfile и обновления UI:
- Обновлен метод fetchUserProfile для обеспечения выполнения всех действий, связанных с пользовательским интерфейсом, и доступа к userProfile в главном потоке.
- Тело обратного вызова обёрнуто в DispatchQueue.main.async для гарантии потокобезопасности, при обновлении userProfile и закрытии индикатора прогресса.
- Решает потенциальные проблемы с многопоточным доступом к свойству userProfile и обеспечивает стабильное поведение пользовательского интерфейса.

vadimrn committed:
Избавление от force unwrapping при обновлении элементов пользовательского интерфейса в методе updateUI, предотвращая краш приложения при потере контекста
self

vadimrn committed:
Улучшение читаемости кода путём ограничения длины строк:
- Отрефакторил метод updateUI, разделив длинные строки на более короткие для лучшей читаемости.
- Обновлены строки с созданием массива UI элементов, чтобы они соответствовали ограничению в 100-120 символов.

vadimrn committed:
Ссылка на командный таск-трекер:
https://github.com/users/Kolesnikov-Eugene/projects/2
Архитектура проекта: MVVM
Верстка UI: в коде
Эпик: Профиль
1/3 эпика согласно плану

vadimrn committed:
Merge branch 'vadinur/profile-1-3' of https://github.com/Kolesnikov-Eugene/iOS-FakeNFT-StarterProject-Public into vadinur/profile-1-3

vadimrn committed:
Revert "добавлены цвета"

This reverts commit 83e087fb9dd6807746b10cf16307abece6f3cb0f.

vadimrn committed:
Revert "добавлены цвета"

This reverts commit 83e087fb9dd6807746b10cf16307abece6f3cb0f.

vadimrn committed:
Revert "Revert "добавлены цвета""

This reverts commit 849417b3e27f307381ed84f73d3a42abc4046e6d.

vadimrn committed:
Revert "Revert "Revert "добавлены цвета"""

This reverts commit 84aa7e6758086485ccdc28ef3d7fbdddb4e65688.

vadimrn committed:
Revert "Revert "Revert "Revert "добавлены цвета""""

This reverts commit 2763c480f918244eb4ff709b8186f88298211ad5.

vadimrn committed:
Добавлена обработка ошибок при загрузке изображения

DanilaG commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,134 @@
+                    self.userProfile = userProfile
+                case .failure(let error):
+                    // ToDo: - Уведомление об ошибке
+                    print(error)
```
Давай добавим тут тоже отображение алерта об сетевой ошибке

vadimrn replied:
исправил

DanilaG commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,134 @@
+                
+            }
+
+                // ToDo: - Уведомление об ошибке
```
Давай уберём комментарий или добавим другой вид обработки ошибок)

vadimrn replied:
исправил

DanilaG reviewed (commented):
Привет! Осталось совсем немного, 2 небольшие правки:
- удалить коммент/показать алерт
- показать алерт (как-нибудь по другому обработать сетевую ошибку)

vadimrn committed:
Добавлена обработка ошибок для fetchProfile и updateProfil

DanilaG commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -42,7 +42,8 @@ final class ProfileViewModel: ProfileViewModelProtocol {
                 case .failure(let error):
-                    // ToDo: - Уведомление об ошибке
+                    print(error)
+                    ProgressHUD.showError("Ошибка загрузки профиля")
```
Круто, что добавлено отображение алерта, однако пользователь его не увидет( Можно попробовать проверить, запустив приложение без интернета. 
 
Это связанно с особеностью работы бибилиотеки. Чтобы это исправить,  давай переместим `ProgressHUD.dismiss()` с 39 строчке на строку после `case .success(let userProfile)` и перед `self.userProfile = userProfile`. Будет выглядить вот так:
```
            DispatchQueue.main.async {
                guard let self = self else { return }
                
                switch result {
                case .success(let userProfile):
                    ProgressHUD.dismiss()
                    self.userProfile = userProfile
                case .failure(let error):
                    print(error)
                    ProgressHUD.showError("Ошибка загрузки профиля")
                    print(error)
                }
            }
```

Перед отправкой, проверь, пожалуйста, что сообщение показалось при выключенном интернете. При этом, оно скроется через несколько секунд (это ок, это такой компонент)

vadimrn replied:
Исправил

vadimrn committed:
исправлена обработка результат асинхронной операции, обновляя пользовательский интерфейс в соответствии с результатом
